validity-test-ospath/OsPath/Gen/Include.hs

```diff
@@ -0,0 +1,145 @@
+  genValid = (Path . ([OsString.pstr|/|] <>) . (<> OsString.singleton OsPath.pathSeparator) <$> genValid) `suchThat` isValid
+  shrinkValid = filter isValid . shrinkValidWith parseAbsDir
+
+instance GenValid (Path Rel File) where
+  genValid = (Path <$> genValid) `suchThat` isValid
+  shrinkValid = filter isValid . shrinkValidWith parseRelFile
+
+instance GenValid (Path Rel Dir) where
+  genValid = (Path . (<> OsString.singleton OsPath.pathSeparator) <$> genValid) `suchThat` isValid
+  shrinkValid = filter isValid . shrinkValidWith parseRelDir
+
+instance GenValid (SomeBase Dir) where
+  genValid = genValidStructurallyWithoutExtraChecking
+  shrinkValid = shrinkValidStructurallyWithoutExtraFiltering
+
+instance GenValid (SomeBase File) where
+  genValid = genValidStructurallyWithoutExtraChecking
+  shrinkValid = shrinkValidStructurallyWithoutExtraFiltering
+
+validateCommon :: Path b t -> Validation
+validateCommon (Path fp) = mconcat
+  [ declare "System.FilePath considers the path valid if it's not empty." $
+      OsPath.isValid fp || OsString.null fp
+  , declare "The path does not contain a '..' path component." $
+      not (hasParentDir fp)
+  ]
+
+validateDirectory :: Path b Dir -> Validation
+validateDirectory (Path fp) = mconcat
+  [ declare "The path has a trailing path separator if it's not empty." $
+      OsPath.hasTrailingPathSeparator fp || OsString.null fp
+  ]
+
+validateFile :: Path b File -> Validation
+validateFile (Path fp) = mconcat
+  [ declare "The path has no trailing path separator." $
+      not (OsPath.hasTrailingPathSeparator fp)
+  , declare "The path does not equal \".\"" $
+      fp /= [OsString.pstr|.|]
+  , declare "The path does not end in /." $
+      not ([OsString.pstr|/.|] `OsString.isSuffixOf` fp)
+  ]
+
+validateAbs :: Path Abs t -> Validation
+validateAbs (Path fp) = mconcat
+  [ declare "The path is absolute." $
+      OsPath.isAbsolute fp
+  ]
+
+validateRel :: Path Rel t -> Validation
+validateRel (Path fp) = mconcat
+  [ declare "The path is relative." $
+      OsPath.isRelative fp
+  ]
+
+shrinkValidWith :: (PLATFORM_PATH -> Maybe (Path a b)) -> Path a b -> [Path a b]
+shrinkValidWith fun (Path f) = filter (/= Path f) . mapMaybe fun $ shrinkValid f
+
+--------------------------------------------------------------------------------
+-- Orphan instances
+
+-- | Generates 'PLATFORM_PATH with a high occurence of @'.'@, @'\/'@ and
+-- @'\\'@ characters. The resulting 'FilePath's are not guaranteed to
+-- be valid.
+instance GenValid PLATFORM_PATH where
+    -- We also need to exclude UTF-16 surrogates.
+    genValid = mconcat <$> listOf (OsString.unsafeEncodeUtf . (:[]) . chr <$> frequency
+        [ (2, choose (0x0, 0xD800 - 1))
+        , (2, choose (0xDFFF + 1, 0x10FFFF))
+        , (1, elements (map ord "./\\"))
+        ]
+        )
+    shrinkValid _ = [] -- TODO: Not yet implemented
+
+instance Validity PLATFORM_PATH where
+    validate = trivialValidation -- TODO: Not yet implemented
```

Comment:
This definitely looks incomplete. I'd be surprised if these were trivially valid.

Reply:
TBH, I simply don't know (yet) what constitutes a _valid_ `PosixPath`/`WindowsPath`... See also my comments in the previous thread.
@hasufell Do you have any idea what a sensible notion would be?

Reply:
Depends what we mean with "valid". Do you mean the `isValid` function from filepath? It's a bit fuzzy. The only thing you need to keep in mind additionally for the new OsPath API when dealing with the constructors yourself is that windows paths are expected to be `[Word16]`. So if you create a one-byte `WindowsString` manually, almost all functions will throw `error`.

Reply:
I think what I meant was answered here: https://stackoverflow.com/questions/1976007/what-characters-are-forbidden-in-windows-and-linux-directory-names

This is not a problem for the instance in the test-suite of this library since we do not pass the generated paths to any OS API, but if we move the instance to a more general-purpose service package like `genvalidity-filepath` we have to implement those restrictions.

Reply:
As some of those SO answers say: it's almost impossible to have an exhaustive list of invalid filepaths on windows. The filepath `isValid` is just best effort.